data/zmod.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2018 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Chris Hughes
+-/
+import data.int.modeq data.fintype data.nat.prime data.nat.gcd
+
+class pos_nat (n : ℕ) := (pos : 0 < n)
+
+attribute [class] nat.prime
```

Comment:
these classes seem like a bad idea

Reply:
I agree, but I couldn't think of an alternative solution. I could define `comm_ring (zmod (succ n))` or demanding a `pnat` is another idea. I could also define `zmod` without `fin` as something like `{a : nat // n = 0 or a < n}`, that way it is completely isomorphic to the quotient. I can't think of a neat solution. The prime instance just seems unavoidable to me.

Reply:
I think `zmod` should take a pnat. For Z/pZ, you can have a def which is not an instance with the inverse function, and have a separate definition `zmodp (p) (pp : is_prime p)` which depends on the primality proof, where you can have the field instance.

Adding the integers as Z/0Z would also solve this problem. Note that your subtype is not correct, since it reduces to `nat` not `int` in the zero case. I haven't ever used a dependent type like this, but you could define it dependently:
```
def zmod : nat -> Type
| 0 := int
| (succ n) := fin (succ n)
```

Reply:
The downside to that is that all the proofs for Z/nZ don't apply to Z/pZ any more so you have to have twice the lemmas?

Reply:
the stuff like typeclasses are easily transferred. There are only a few big theorems that aren't about typeclasses, and these are also easily transferred by defeq.

Reply:
Updated to use pnats.